datastructs/ring.c

```diff
@@ -0,0 +1,76 @@
+// ring.c : 
+/* A ring buffer for multi-threaded programs. A single producer pushes
+* objects into the ring, while a single consumer pops them off for
+* processing. If the ring is full (i.e., the consumer is slower than the
+* producer), then the push operation fails.
+*/
+
+
+
+
+#include <stdlib.h>
+#include <stdio.h>
+
+#include "ring.h"
+
+
+
+
+/* Create a ring buffer with the specified size. Return the ring or NULL
+if there is not enough memory to create. */
+struct ring* ring_create(int size)
+{
+	if (size <= 0)  return NULL;
+
+
+
+	struct ring *rb = (struct ring*) malloc( sizeof(struct ring) );
+	rb->push = 0;
+	rb->pop = 0;
+
+	rb->size = size;
+	rb->count = 0;
+
+	rb->buffer = (void*) malloc( size*sizeof(int) );
+
+	return rb;
+};
+
+/* Add an entry to the ring.
+Return 0 on success, or a sensible error code if ring is full
+*/
+int ring_push(struct ring *rb, void* data)
+{ 
+	if(rb->count >= rb->size)
+		return -1;
+	else
+	{
+		*( (int*)( rb->buffer ) + rb->push ) = *(int*) data;
+		rb->count = rb->count + 1;
+		rb->push = (rb->push +1 ) % rb->size;
+		return 0;
+	}
+
+}
+
+/* Remove an entry from the ring.
+Return a pointer to the data, or NULL if empty
+*/
+
+
+void* ring_pop(struct ring *rb)
+{	
+	void *data= (void*) malloc( sizeof(int) );
```

Comment:
you don't need to be `malloc`ing any memory here---you just want data to be a pointer.  It should point to something inside your rb->buffer, not some new memory you reserve using malloc.